<a href="https://colab.research.google.com/github/PercyAyimbilaNsolemna/Deep-Learning/blob/main/Titanic_Competition/titanic_neural_networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **KAGGLE TITANIC CHALLENGE SOLUTION**

Implementation of kaggle titanic challenge using Deep Learning (Neural Networks)

## **IMPORTING LIBRARIES**

This section imports all the libraries that will be used in building the Machine Learning workflow

In [392]:
#Imports numpy as np
import numpy as np
#Imports pandas as pd
import pandas as pd
#Imports tensorflow as tf
import tensorflow as tf
#Imports train_test_split from scikit learn model_selection
from sklearn.model_selection import train_test_split
#Imports StandardScaler from scikit learn preprocessing
from sklearn.preprocessing import StandardScaler
#Imports mean squared error from scikit learn metrics
from sklearn.metrics import mean_squared_error
#Imports accuracy score from scikit learn metrics
from sklearn.metrics import accuracy_score
#Imports Sequential model from tensorflow keras model
from tensorflow.keras.models import Sequential
#Imports the Dense layer from tensorflow keras layers
from tensorflow.keras.layers import Dense
#Imports OneHotEncode from sklearn preprocessing
from sklearn.preprocessing import OneHotEncoder
#Imports Logistic Regression from sklear.linear_model
from sklearn.linear_model import LogisticRegression
#Imports the RandomForestClassifier from sklearn.ensemble
from sklearn.ensemble import RandomForestClassifier
#Imports warning and supress all future warnings
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

#np.random.seed(1) is used to keep all the random function calls consistent.
np.random.seed(1)

# **IMPORTING GOOGLE DRIVE**

This section imports google drive into the notebook so that the dataset for the project can be accessed for training, cross-validation and testing

In [393]:
#Imports drive from google colab
from google.colab import drive

In [394]:
#Connects or mounts google drive to the google colab
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## **TITANIC DATASET**

The titanic dataset will be used to build the neural network. Detailed information on the project can be found on [kaggle](https://www.kaggle.com/c/titanic)

The dataset used can be downloaded and explored [here](https://www.kaggle.com/c/titanic/data)

In [395]:
#Loads the training set
X = pd.read_csv('/content/drive/MyDrive/Datasets/titanic/train.csv')

In [396]:
#Loads the test set
X_test = pd.read_csv('/content/drive/MyDrive/Datasets/titanic/test.csv')

In [397]:
#Loads the sample gender submission file, gives a sample look of how the outcomes should be  submitted
gender_submission = pd.read_csv('/content/drive/MyDrive/Datasets/titanic/gender_submission.csv')

## **GENDER SUBMISSION VISUALIZATION**

This section gives a snapshot of the outcome of the model and the look of the file that is supposed to be submitted.

In [398]:
#Outputs the type of data been used
print(f'The type of the gender submission is \n{type(gender_submission)}')

The type of the gender submission is 
<class 'pandas.core.frame.DataFrame'>


In [399]:
#Outputs a detailed information of the gender submission dataset
gender_submission.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   PassengerId  418 non-null    int64
 1   Survived     418 non-null    int64
dtypes: int64(2)
memory usage: 6.7 KB


In [400]:
#Outputs the shape of the gender submission dataset
print(f'The shape of the gender submission is {gender_submission.shape}')

The shape of the gender submission is (418, 2)


In [401]:
#Outputs the titles of the two columns
print(f'The titles of the gender submission file are: \n{gender_submission.columns}')

The titles of the gender submission file are: 
Index(['PassengerId', 'Survived'], dtype='object')


In [402]:
#Outputs the head of the gender_submission file
gender_submission.head()

PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1

In [403]:
#Outputs the tail of the gender submission dataset
gender_submission.tail()

PassengerId  Survived
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

## **INSIGHTS AFTER DATA EXPLORATION**

This section gives a concise information gained from the gender submission Data Ecploration.

The **predicted values (y^)** should be submitted using two columns that is the **passengerID** and the survival state, **whether he/she survived (1) or did not survive (0)**

## **DATA VISUALIZATION (TRAIN DATASET)**

In this section we will focus on visualizing the train dataset. The major areas that will be checked are checking for any row with null cell and deleting the cell(s), checking the data types stored by each feature and performing data type convertion if the need arises, and one hot encoding.

In [404]:
#Outputs the data type of the train dataset
print(f'The data type of the train dataset is: \n{type(X)}')

The data type of the train dataset is: 
<class 'pandas.core.frame.DataFrame'>


In [405]:
#Outputs information about the train dataset
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [406]:
#Outputs the shape of the train dataset
print(f'The shape of the train dataset is: {X.shape}')

The shape of the train dataset is: (891, 12)


In [407]:
#Outputs the features of the train dataset
print(f'The features in the train dataset are: \n{X.columns}')

The features in the train dataset are: 
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


## **FEATURES DEFINITION**

This section throws light on the features of the train dataset

 **VARIABLE**      | **DEFINITION**   | **KEY**
-------------------|------------------|----------------
Survived           | Survival         | 0 = No, 1 = Yes
Pclass             | Ticket class     | 1st, 2nd or 3rd
Name               | Name of Passenger |
PassengerID        | Pasenger ID       |
Sex                | Sex              |
Age                | Age in years     |
SibSp              | # of siblings / spouses aboard the Titanic |
Parch              | # of parents / children aboard the Titanic |
Ticket             | Ticket number    |
Fare               | Passenger fare   |
Cabin              | Cabin number     |
Embarked           | Port of Embarkation | C = Cherbourg, Q = Queenstown, Southampton

In [408]:
#Outputs the head of the train dataset
X.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [409]:
#Outputs the tail of the train dataset
X.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

## **EMPTY CELLS IN TRAIN AND TEST SET**

This section reviews the train and test dataset for NaN cells and fills them up.

The **Cabin** feature will be dropped due to the large number of empty cells

In [410]:
#Iterates through all the features and outputs the length of the features with NaN values in the training set
print('Number of NaN cells in each feature in the training set \n')
for feature in X.columns:
 print(f'The {feature} column has {X[pd.isna(X[feature])].shape[0]} cells with NaN value(s)')


Number of NaN cells in each feature in the training set 

The PassengerId column has 0 cells with NaN value(s)
The Survived column has 0 cells with NaN value(s)
The Pclass column has 0 cells with NaN value(s)
The Name column has 0 cells with NaN value(s)
The Sex column has 0 cells with NaN value(s)
The Age column has 177 cells with NaN value(s)
The SibSp column has 0 cells with NaN value(s)
The Parch column has 0 cells with NaN value(s)
The Ticket column has 0 cells with NaN value(s)
The Fare column has 0 cells with NaN value(s)
The Cabin column has 687 cells with NaN value(s)
The Embarked column has 2 cells with NaN value(s)


The **Age feature** has **177 training examples** with NaN values

The **Cabin feature** has **687 training examples** with NaN values

The **Embarked feature** has **2 training examples** with NaN values

In [411]:
#Loops through all the features and outputs the length of the feature with NaN values in the test set
print('The number of NaN cells in each feature in the test set \n')
for feature in X_test.columns:
  print(f'The {feature} column has {X_test[pd.isna(X_test[feature])].shape[0]} cells with NaN value(s)')

The number of NaN cells in each feature in the test set 

The PassengerId column has 0 cells with NaN value(s)
The Pclass column has 0 cells with NaN value(s)
The Name column has 0 cells with NaN value(s)
The Sex column has 0 cells with NaN value(s)
The Age column has 86 cells with NaN value(s)
The SibSp column has 0 cells with NaN value(s)
The Parch column has 0 cells with NaN value(s)
The Ticket column has 0 cells with NaN value(s)
The Fare column has 1 cells with NaN value(s)
The Cabin column has 327 cells with NaN value(s)
The Embarked column has 0 cells with NaN value(s)


The **Age feature** has 86 cells with NaN values

The **Cabin feature** has 327 cells with NaN values

The **Fare feature** has 1 cells with NaN values

In [412]:
#Fills all NaN cells in the Age feature with the median age to prevent bias in the train and test set
X['Age'].fillna(X['Age'].median(), inplace=True)
X_test['Age'].fillna(X_test['Age'].median(), inplace=True)

#Fills all NaN cells in the Cabin feature with the mode of the Embarked feature to maintain the distribution of the train
X['Embarked'].fillna(X['Embarked'].mode()[0], inplace=True)

#Fills all NaN cells in the Cabin feature with the mode of the Embarked feature to maintain the distribution of the train and test set
X_test['Fare'].fillna(X_test['Fare'].median(), inplace=True)

In [413]:
#Loops through the training dataset to check is there is an NaN value in any of the cells excluding the Cabin feature
print('Checking for NaN value(s) in each of the features in the training set: \n')

for feature in X.columns:
  print(f'The {feature} column has {X[pd.isna(X[feature])].shape[0]} cells with NaN value(s)')

Checking for NaN value(s) in each of the features in the training set: 

The PassengerId column has 0 cells with NaN value(s)
The Survived column has 0 cells with NaN value(s)
The Pclass column has 0 cells with NaN value(s)
The Name column has 0 cells with NaN value(s)
The Sex column has 0 cells with NaN value(s)
The Age column has 0 cells with NaN value(s)
The SibSp column has 0 cells with NaN value(s)
The Parch column has 0 cells with NaN value(s)
The Ticket column has 0 cells with NaN value(s)
The Fare column has 0 cells with NaN value(s)
The Cabin column has 687 cells with NaN value(s)
The Embarked column has 0 cells with NaN value(s)


In [414]:
#Loops through the test set for any cell(s) with an NaN value excluding the Cabin feature
print('Checking for NaN value(s) in each of the features in the test set: \n')

for feature in X_test.columns:
  print(f'The {feature} column has {X_test[pd.isna(X_test[feature])].shape[0]} cells with NaN value(s)')

Checking for NaN value(s) in each of the features in the test set: 

The PassengerId column has 0 cells with NaN value(s)
The Pclass column has 0 cells with NaN value(s)
The Name column has 0 cells with NaN value(s)
The Sex column has 0 cells with NaN value(s)
The Age column has 0 cells with NaN value(s)
The SibSp column has 0 cells with NaN value(s)
The Parch column has 0 cells with NaN value(s)
The Ticket column has 0 cells with NaN value(s)
The Fare column has 0 cells with NaN value(s)
The Cabin column has 327 cells with NaN value(s)
The Embarked column has 0 cells with NaN value(s)


There are no empty cells in both the traiuning and testing dataset excluding the cabin feature. The Cabin feature will be dropped in the next cell.

## **DROPPING THE CABIN FEATURE**

The **Cabin feature** is dropped due to the huge number of missing values or null values in both the train and test dataset

In [415]:
#Drops the Cabin column in both the train and test set
try:
  X.drop('Cabin', axis=1, inplace=True)
  X_test.drop('Cabin', axis=1, inplace=True)

except KeyError:
  print('The Cabin feature has already been dropped')

Drops the PassengerId feature since it will not be used in training the model

In [416]:
try:
  X.drop('PassengerId', axis=1, inplace=True)

except KeyError:
  print('The PassengerId has been dropped')

In [417]:
#Outputs the columns in both the train and test data set
print('The columns in the train set are: ')
print(X.columns)

print('\nThe columns in the test set are: ')
print(X_test.columns)

The columns in the train set are: 
Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
       'Fare', 'Embarked'],
      dtype='object')

The columns in the test set are: 
Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Embarked'],
      dtype='object')


## **FEATURE ENGINEERING**

This section deals with creating new features out of the original features.

The **Sibsp and the Parch features** will be combined to get the **family size**

Aside that the **Age feature** will be divided into four sections that is **Children, Teenage, Adult and Elder.**

The **Fare** will equally be divided into four separate sections that's **Low fare, Median fare, Average fare and high fare.**

The Ticket, Age, Name and Fare feature will be dropped

In [418]:
#Adds the Sibsp and Parch features
#The one added is the person whose parents or siblings onboard the titanic
X['FamilySize'] = X['SibSp'] + X['Parch'] + 1

In [419]:
#Adds the SibSp and the Parch features in the test set
#The one added to the family size represents the person whose family onboard the titanic
X_test['FamilySize'] = X_test['SibSp'] + X_test['Parch'] + 1

In [420]:
#Outputs the least and highest value in the train and test set for splitting the ages
print(f"The least Age in the train  set is {X['Age'].min()} and the highest is {X['Age'].max()}")
print(f"The least Age in the test set is {X_test['Age'].min()} and the highest is {X_test['Age'].max()}")

The least Age in the train  set is 0.42 and the highest is 80.0
The least Age in the test set is 0.17 and the highest is 76.0


In [421]:
#Subdivides the Age feature to Children, Tennagers, Adults and Elders
X['Age_bin'] = pd.cut(X['Age'], bins=[X['Age'].min()-1, 12, 20, 40, X['Age'].max()+1], labels=['Children', 'Teenagers', 'Adults', 'Elders'])
X_test['Age_bin'] = pd.cut(X_test['Age'], bins=[0, 12, 20, 40, 120], labels=['Children', 'Teenagers', 'Adults', 'Elders'])

In [422]:
#Outputs the least and greatest value in the test and train set
print(f"The least fare in the train  set is {X['Fare'].min()} and the highest is {X['Fare'].max()}")
print(f"The least fare in the test set is {X_test['Fare'].min()} and the highest is {X_test['Fare'].max()}")

The least fare in the train  set is 0.0 and the highest is 512.3292
The least fare in the test set is 0.0 and the highest is 512.3292


In [423]:
#Divides the Fare feature into Low Fare, Median Fare, Average Fare and High Fare in the train and test set
X['Fare_bin'] = pd.cut(X['Fare'], bins=[X['Fare'].min()-1, 50, 138, 296, X['Fare'].max()+1], labels=['LowFare', 'MedianFare', 'AverageFare', 'HighFare'])
X_test['Fare_bin'] = pd.cut(X_test['Fare'], bins=[X['Fare'].min()-1, 50, 138, 296, X['Fare'].max()+1], labels=['LowFare', 'MedianFare', 'AverageFare', 'HighFare'])

In [424]:
#Drops the features in the list
drop_features = ['Age', 'Ticket', 'Name', 'Fare']

X.drop(drop_features, axis=1, inplace=True)
X_test.drop(drop_features, axis=1, inplace=True)

## **ONE HOT ENCODING**

In this section, we will dive deep into one hot encoding. One hot encoding is transforming categorical data (text) into numerical data for processing.

In one hot encoding only one of the categorical data in the specified featur is 1 and the rest of the distinct values are set to zero.

For instance the **Age feature** has only two distinct values **Male and Female**. If the person is a Male then the one hot encoded feature will be [1 0]  where the first feature is Male and the second is Female. Also, if a person is Femaele then the one hot encoded feature will be [0 1] where the first column is a Male and the second column is Female.

The features that will be one hot encoded are:
* **Sex**
* **Age_bin**
* **Embarked**
* **Fare_bin**

In [425]:
#Loops through the columns excluding the features in the list and displays the distinct values in each column
for column in X.columns:
    print(f'The distinct values in the \033[1m {column} \033[0;0m column are: ')
    print(f'{X[column].unique()} \n')

The distinct values in the  Survived  column are: 
[0 1] 

The distinct values in the  Pclass  column are: 
[3 1 2] 

The distinct values in the  Sex  column are: 
['male' 'female'] 

The distinct values in the  SibSp  column are: 
[1 0 3 4 2 5 8] 

The distinct values in the  Parch  column are: 
[0 1 2 5 3 4 6] 

The distinct values in the  Embarked  column are: 
['S' 'C' 'Q'] 

The distinct values in the  FamilySize  column are: 
[ 2  1  5  3  7  6  4  8 11] 

The distinct values in the  Age_bin  column are: 
['Adults', 'Elders', 'Children', 'Teenagers']
Categories (4, object): ['Children' < 'Teenagers' < 'Adults' < 'Elders'] 

The distinct values in the  Fare_bin  column are: 
['LowFare', 'MedianFare', 'AverageFare', 'HighFare']
Categories (4, object): ['LowFare' < 'MedianFare' < 'AverageFare' < 'HighFare'] 



In [426]:
#Loops through all the features and outputs the datatypes stored by each feature
print('The datatypes of the features are: \n')
X.dtypes

The datatypes of the features are: 



Survived         int64
Pclass           int64
Sex             object
SibSp            int64
Parch            int64
Embarked        object
FamilySize       int64
Age_bin       category
Fare_bin      category
dtype: object

In [427]:
#Creates an object from the OneHotEncoder class
ohe = OneHotEncoder()

### **Performs one hot encoding on the traning set**

In [428]:
#Uses the fit_transform method in the OneHotEncoder class to convert the categorical data to numerical values
transformed_features_train = ohe.fit_transform(X[X.select_dtypes(include=['object', 'category']).columns]).toarray()
#Converts the transformed features from float to ints
transformed_features_train = transformed_features_train.astype(int)
#Outputs the transformed features
print(transformed_features_train)

[[0 1 0 ... 0 1 0]
 [1 0 1 ... 0 0 1]
 [1 0 0 ... 0 1 0]
 ...
 [1 0 0 ... 0 1 0]
 [0 1 1 ... 0 1 0]
 [0 1 0 ... 0 1 0]]


In [429]:
#Extracts the one hot encoded features
encoded_features_train = ohe.get_feature_names_out()

#Outputs the encoded features
for encoded_feature in encoded_features_train:
  print(encoded_feature)

#Outputs the length of the encoded_features
print(f'\nThe number of features that were encoded are: {len(encoded_features_train)}')

Sex_female
Sex_male
Embarked_C
Embarked_Q
Embarked_S
Age_bin_Adults
Age_bin_Children
Age_bin_Elders
Age_bin_Teenagers
Fare_bin_AverageFare
Fare_bin_HighFare
Fare_bin_LowFare
Fare_bin_MedianFare

The number of features that were encoded are: 13


In [430]:
#Creating a pandas Dadaframe from the one hot encoded features
one_hot_encoded_features_train = pd.DataFrame(transformed_features_train, columns=encoded_features_train)

In [431]:
#Outputs the head of the one_hot_encoded_features
one_hot_encoded_features_train.head()

Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S  Age_bin_Adults  \
0           0         1           0           0           1               1   
1           1         0           1           0           0               1   
2           1         0           0           0           1               1   
3           1         0           0           0           1               1   
4           0         1           0           0           1               1   

   Age_bin_Children  Age_bin_Elders  Age_bin_Teenagers  Fare_bin_AverageFare  \
0                 0               0                  0                     0   
1                 0               0                  0                     0   
2                 0               0                  0                     0   
3                 0               0                  0                     0   
4                 0               0                  0                     0   

   Fare_bin_HighFare  Fare_bin_LowFare  Fare_bin_MedianFare  
0                  0                 1                    0  
1                  0                 0                    1  
2                  0                 1                    0  
3                  0                 0                    1  
4                  0                 1                    0

In [432]:
#Outputs the tail of the one_hot_encoded_features
one_hot_encoded_features_train.tail()

Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S  Age_bin_Adults  \
886           0         1           0           0           1               1   
887           1         0           0           0           1               0   
888           1         0           0           0           1               1   
889           0         1           1           0           0               1   
890           0         1           0           1           0               1   

     Age_bin_Children  Age_bin_Elders  Age_bin_Teenagers  \
886                 0               0                  0   
887                 0               0                  1   
888                 0               0                  0   
889                 0               0                  0   
890                 0               0                  0   

     Fare_bin_AverageFare  Fare_bin_HighFare  Fare_bin_LowFare  \
886                     0                  0                 1   
887                     0                  0                 1   
888                     0                  0                 1   
889                     0                  0                 1   
890                     0                  0                 1   

     Fare_bin_MedianFare  
886                    0  
887                    0  
888                    0  
889                    0  
890                    0

In [433]:
#Outputs the shape of the one_hot_encoded_features to be sure the shape tallies with the training data
print(f'The shape of the one hot encoded features is {one_hot_encoded_features_train.shape}')

The shape of the one hot encoded features is (891, 13)


### **Performs one hot encoding on the test set**

In [434]:
#Uses the fit_transform method in the OneHotEncoder class to convert the categorical data to numerical values in the test set
transformed_features_test = ohe.fit_transform(X_test[X_test.select_dtypes(include=['object', 'category']).columns]).toarray()
#Converts the transformed features from float to ints
transformed_features_test = transformed_features_test.astype(int)
#Outputs the transformed features
print(transformed_features_test)

[[0 1 0 ... 0 1 0]
 [1 0 0 ... 0 1 0]
 [0 1 0 ... 0 1 0]
 ...
 [0 1 0 ... 0 1 0]
 [0 1 0 ... 0 1 0]
 [0 1 1 ... 0 1 0]]


In [435]:
#Extracts the one hot encoded features
encoded_features_test = ohe.get_feature_names_out()

In [436]:
#Outputs the names of the encoded features in the test_set
for encoded_feature in encoded_features_test:
  print(encoded_feature)

Sex_female
Sex_male
Embarked_C
Embarked_Q
Embarked_S
Age_bin_Adults
Age_bin_Children
Age_bin_Elders
Age_bin_Teenagers
Fare_bin_AverageFare
Fare_bin_HighFare
Fare_bin_LowFare
Fare_bin_MedianFare


In [437]:
#Convert the test set one hot encoded features to a pandas dataFrame
one_hot_encoded_features_test = pd.DataFrame(transformed_features_test, columns=encoded_features_test)

In [438]:
#Ouputs the head of the one hot encoded features for the test set
one_hot_encoded_features_test.head()

Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S  Age_bin_Adults  \
0           0         1           0           1           0               1   
1           1         0           0           0           1               0   
2           0         1           0           1           0               0   
3           0         1           0           0           1               1   
4           1         0           0           0           1               1   

   Age_bin_Children  Age_bin_Elders  Age_bin_Teenagers  Fare_bin_AverageFare  \
0                 0               0                  0                     0   
1                 0               1                  0                     0   
2                 0               1                  0                     0   
3                 0               0                  0                     0   
4                 0               0                  0                     0   

   Fare_bin_HighFare  Fare_bin_LowFare  Fare_bin_MedianFare  
0                  0                 1                    0  
1                  0                 1                    0  
2                  0                 1                    0  
3                  0                 1                    0  
4                  0                 1                    0

In [439]:
one_hot_encoded_features_test.tail()

Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S  Age_bin_Adults  \
413           0         1           0           0           1               1   
414           1         0           1           0           0               1   
415           0         1           0           0           1               1   
416           0         1           0           0           1               1   
417           0         1           1           0           0               1   

     Age_bin_Children  Age_bin_Elders  Age_bin_Teenagers  \
413                 0               0                  0   
414                 0               0                  0   
415                 0               0                  0   
416                 0               0                  0   
417                 0               0                  0   

     Fare_bin_AverageFare  Fare_bin_HighFare  Fare_bin_LowFare  \
413                     0                  0                 1   
414                     0                  0                 0   
415                     0                  0                 1   
416                     0                  0                 1   
417                     0                  0                 1   

     Fare_bin_MedianFare  
413                    0  
414                    1  
415                    0  
416                    0  
417                    0

## **MERGE THE ONE HOT ENCODED FEATURES TO THE DATASETS**

In this section the one hot encoded features will be merged with the respective dataset.

### Reset the indices in the train and one hot encoded features train

In [440]:
#Reset indices to ensure alignemnt before merging to prevent NA in some cells
X.reset_index(drop=True, inplace=True)
one_hot_encoded_features_train.reset_index(drop=True, inplace=True)

#Outputs the shapes to be sure they are still intact
print(f'The shape of X is {X.shape}')
print(f'The shape of one hot encoded features train is {one_hot_encoded_features_train.shape}')

The shape of X is (891, 9)
The shape of one hot encoded features train is (891, 13)


### Concatenates the X and one hot encoding features train

In [441]:
#Concatenates X and one_hot_encoded_features_train dataframes
X_meg = pd.concat([X, one_hot_encoded_features_train], axis=1)

#Outputs the concatenated dataframe
X_meg

Survived  Pclass     Sex  SibSp  Parch Embarked  FamilySize    Age_bin  \
0           0       3    male      1      0        S           2     Adults   
1           1       1  female      1      0        C           2     Adults   
2           1       3  female      0      0        S           1     Adults   
3           1       1  female      1      0        S           2     Adults   
4           0       3    male      0      0        S           1     Adults   
..        ...     ...     ...    ...    ...      ...         ...        ...   
886         0       2    male      0      0        S           1     Adults   
887         1       1  female      0      0        S           1  Teenagers   
888         0       3  female      1      2        S           4     Adults   
889         1       1    male      0      0        C           1     Adults   
890         0       3    male      0      0        Q           1     Adults   

       Fare_bin  Sex_female  ...  Embarked_Q  Embarked_S  Age_bin_Adults  \
0       LowFare           0  ...           0           1               1   
1    MedianFare           1  ...           0           0               1   
2       LowFare           1  ...           0           1               1   
3    MedianFare           1  ...           0           1               1   
4       LowFare           0  ...           0           1               1   
..          ...         ...  ...         ...         ...             ...   
886     LowFare           0  ...           0           1               1   
887     LowFare           1  ...           0           1               0   
888     LowFare           1  ...           0           1               1   
889     LowFare           0  ...           0           0               1   
890     LowFare           0  ...           1           0               1   

     Age_bin_Children  Age_bin_Elders  Age_bin_Teenagers  \
0                   0               0                  0   
1                   0               0                  0   
2                   0               0                  0   
3                   0               0                  0   
4                   0               0                  0   
..                ...             ...                ...   
886                 0               0                  0   
887                 0               0                  1   
888                 0               0                  0   
889                 0               0                  0   
890                 0               0                  0   

     Fare_bin_AverageFare  Fare_bin_HighFare  Fare_bin_LowFare  \
0                       0                  0                 1   
1                       0                  0                 0   
2                       0                  0                 1   
3                       0                  0                 0   
4                       0                  0                 1   
..                    ...                ...               ...   
886                     0                  0                 1   
887                     0                  0                 1   
888                     0                  0                 1   
889                     0                  0                 1   
890                     0                  0                 1   

     Fare_bin_MedianFare  
0                      0  
1                      1  
2                      0  
3                      1  
4                      0  
..                   ...  
886                    0  
887                    0  
888                    0  
889                    0  
890                    0  

[891 rows x 22 columns]

In [442]:
#Outputs info of the concatenated dataframe
print(X_meg.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   Survived              891 non-null    int64   
 1   Pclass                891 non-null    int64   
 2   Sex                   891 non-null    object  
 3   SibSp                 891 non-null    int64   
 4   Parch                 891 non-null    int64   
 5   Embarked              891 non-null    object  
 6   FamilySize            891 non-null    int64   
 7   Age_bin               891 non-null    category
 8   Fare_bin              891 non-null    category
 9   Sex_female            891 non-null    int64   
 10  Sex_male              891 non-null    int64   
 11  Embarked_C            891 non-null    int64   
 12  Embarked_Q            891 non-null    int64   
 13  Embarked_S            891 non-null    int64   
 14  Age_bin_Adults        891 non-null    int64   
 15  Age_bi

### Resets the X_test and the oen hot encoded features

In [443]:
#Resets indices to ensure alignment before merging the dataframes
X_test.reset_index(drop=True, inplace=True)
one_hot_encoded_features_test.reset_index(drop=True, inplace=True)

#Outputs the shape to be sure they are the same
print(f'The shape of the X_test is {X_test.shape}')
print(f'The shape of the one hot encoded features test is {one_hot_encoded_features_test.shape}')

The shape of the X_test is (418, 9)
The shape of the one hot encoded features test is (418, 13)


### Concatenates the X_test and the one hot encoded features test

In [444]:
X_test_meg = pd.concat([X_test, one_hot_encoded_features_test], axis=1)

In [445]:
#Outputs detailed info of the merged dataframe
print(X_test_meg.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   PassengerId           418 non-null    int64   
 1   Pclass                418 non-null    int64   
 2   Sex                   418 non-null    object  
 3   SibSp                 418 non-null    int64   
 4   Parch                 418 non-null    int64   
 5   Embarked              418 non-null    object  
 6   FamilySize            418 non-null    int64   
 7   Age_bin               418 non-null    category
 8   Fare_bin              418 non-null    category
 9   Sex_female            418 non-null    int64   
 10  Sex_male              418 non-null    int64   
 11  Embarked_C            418 non-null    int64   
 12  Embarked_Q            418 non-null    int64   
 13  Embarked_S            418 non-null    int64   
 14  Age_bin_Adults        418 non-null    int64   
 15  Age_bi

## **ONE HOT ENCODED FEATURES EXPLORATION**

This section of the notebook delves into the features that will be deleted, renamed or retained.

**Since the Sex feature is binary which is Sex_female and Sex_male, the two features will be combined with a male being 1 and a female 0.**

The Sex_female and sex feature will be dropped and the Sex_male feature will be renamed as Sex.

The encoded features for **Embarked, Age and Fare** will be retained with no modifications to the feature names

**Features to drop**
- Sex
- Embarked
- Sex_female
- Age_bin
- Fare_bin

**FEatures to rename**
- Sex_male to Sex
- Age_bin_Adults to Age_Adults
- Age_bin_Childrerd to Age_Children
- Age_bin_Elders to Age_Elders
- Age_bin_Teenagers to Age_Teenagers
- Fare_bin_AverageFare to Fare_Average
- Fare_bin_HighFare to Fare_High
- Fare_bin_LowFare to Fare_Low
- Fare_bin_MedianFare to Fare_Median

The same idea will be applied to both one hot encoded features for the train and test set

### DROPS FEATURES IN THE TRAIN AND TEST SET

In [446]:
features_to_drop = ['Sex', 'Embarked', 'Sex_female', 'Age_bin', 'Fare_bin']

In [447]:
#Drops the features in the features to drop from the X_new
try:
  X_meg.drop(features_to_drop, axis=1, inplace=True)

except KeyError:
  print('The features to drop have been dropped already')

In [448]:
#Drops the features in the features to drop from the X_test_meg
try:
  X_test_meg.drop(features_to_drop, axis=1, inplace=True)

except KeyError:
  print('The features to drop have already been dropped')

### RENAME FEATURES IN THE TRAIN SET

In [449]:
#Features to rename
features_to_rename = {'Sex_male': 'Sex', 'Age_bin_Adults': 'Age_Adults', 'Age_bin_Children': 'Age_Children', 'Age_bin_Elders': 'Age_Elders',
                      'Age_bin_Teenagers': 'Age_Teenagers', 'Fare_bin_AverageFare': 'Fare_Average', 'Fare_bin_HighFare': 'Fare_High',
                      'Fare_bin_LowFare': 'Fare_Low', 'Fare_bin_MedianFare': 'Fare_Median'
                      }

# Sex_male to Sex
# Age_bin_Adults to Age_Adults
# Age_bin_Childrerd to Age_Children
# Age_bin_Elders to Age_Elders
# Age_bin_Teenagers to Age_Teenagers
# Fare_bin_AverageFare to Fare_Average
# Fare_bin_HighFare to Fare_High
# Fare_bin_LowFare to Fare_Low
# Fare_bin_MedianFare to Fare_Median

In [450]:
X_meg.rename(columns=features_to_rename, inplace=True)

In [451]:
for column in X_meg.columns:
  print(column)

Survived
Pclass
SibSp
Parch
FamilySize
Sex
Embarked_C
Embarked_Q
Embarked_S
Age_Adults
Age_Children
Age_Elders
Age_Teenagers
Fare_Average
Fare_High
Fare_Low
Fare_Median


### RENAME FEATURES IN THE TEST SET

In [452]:
#Renames the features in the merged test set
X_test_meg.rename(columns=features_to_rename, inplace=True)

In [453]:
for column in X_test_meg.columns:
  print(column)

PassengerId
Pclass
SibSp
Parch
FamilySize
Sex
Embarked_C
Embarked_Q
Embarked_S
Age_Adults
Age_Children
Age_Elders
Age_Teenagers
Fare_Average
Fare_High
Fare_Low
Fare_Median


## **SPLITTING DATA INTO TRAIN AND TARGET VALUES**

This section of the notebook focuses on splitting data in the X_meg to features and target values since this is a supervised learning problem

The survived feature will be our target value since it has the overall outcome whether the passenger survived or not.

All the other features will be features

In [454]:
#Extracts all the features excluding the Survived feature and converts it to a numpy array into the X_train_unscaled
X_train_orig_unscaled = X_meg.drop(['Survived'], axis=1).values
print(f'The shape of the X_train_unscaled is {X_train_orig_unscaled.shape}\n')
print(type(X_train_orig_unscaled))

#Puts the survived feature in y which is the target value
y = X_meg['Survived'].values
print(f'\nThe shape of the target value y is {y.shape}')

The shape of the X_train_unscaled is (891, 16)

<class 'numpy.ndarray'>

The shape of the target value y is (891,)


In [455]:
#Drops the PassengerId and converts the pandas dataframe to a numpy array
X_test_unscaled = X_test_meg.drop(['PassengerId'], axis=1).values
print(f'The shape of the X_test_unscaled is {X_test_unscaled.shape}')

The shape of the X_test_unscaled is (418, 16)


## **SPLITTING THE TRAINING DATA**

The training data will be splitted into train and cross-validation set.

90% of the training data will be used to train the model and 10% of the training data will be used to validate the model to access its performance.

In [456]:
#Splits the X_train_unscaled data into training and cross-validation set
#90% of the X_train_unscaled data will be used for training whiles 10% will be used to assess the performance of the model
X_train_unscaled, X_cv_unscaled, y_train, y_cv = train_test_split(X_train_orig_unscaled, y, train_size=0.90, random_state=42, stratify=y)


print(f'The shape of the training set (input) is {X_train_unscaled.shape}')
print(f'The shape of the training set (target) is {y_train.shape}\n')
print(f'The shape of the cross validation set (input) is {X_cv_unscaled.shape}')
print(f'The shape of the cross validation set (target) is {y_cv.shape}')

The shape of the training set (input) is (801, 16)
The shape of the training set (target) is (801,)

The shape of the cross validation set (input) is (90, 16)
The shape of the cross validation set (target) is (90,)


## **FEATURE SCALING**

Now all the features are numerical ready to be used to train the model.

But there is one thing at hand, our data is not balanced. Using an unscaled data to train a model makes gradient descent converge slowly.

In this section, we will scale our data both the training and test set to ensure that all the features have values between 0 and 1.

This makes gradient descent run faster and prevent overfitting.

In [457]:
#Instantiate the standard scaler from scikit learn
standardScaler = StandardScaler()

#Scales the training set
X_train_scaled = standardScaler.fit_transform(X_train_unscaled)

#Outputs the calculated mean by standard scaler
print(f'The calculated mean for the training data is {standardScaler.mean_}\n')

#Outputs the minimum and maximum value in the X_train_unscaled dataset
print(f'The minimum value in the X_train_unscaled is {X_train_unscaled.min()}')
print(f'The maximum value in the X_train_unscaled is {X_train_unscaled.max()}\n')

#Outputs the minimum and maximum value in the X_train_scaled
print(f'The minimum value in the X_train_scaled is {X_train_scaled.min()}')
print(f'The maximum value in the X_train_scaled is {X_train_scaled.max()}')

The calculated mean for the training data is [2.3133583  0.51935081 0.39450687 1.91385768 0.64918851 0.19101124
 0.082397   0.72659176 0.63171036 0.07990012 0.16729089 0.12109863
 0.03121099 0.00249688 0.82022472 0.14606742]

The minimum value in the X_train_unscaled is 0
The maximum value in the X_train_unscaled is 11

The minimum value in the X_train_scaled is -2.1360009363293746
The maximum value in the X_train_scaled is 19.987496091306557


In [458]:
#Scales the X_cv_unscaled using the mean and standard deviation used to scale the training data
X_cv_scaled = standardScaler.fit_transform(X_cv_unscaled)

#Scales teh X_test_unscaled using the mean and standard deviation used to scale the training data
X_test_scaled = standardScaler.fit_transform(X_test_unscaled)

## **DEVELOPING THE MODEL**

Everything is now set to develop the model. **Deep Learning algorithms** will be used to develop the model.

The **linear, relu, tanh and sigmoid activation functions** will be used to train the model and the one that performs better will be used to make prediction on the test set

In [459]:
#A dictionary to store the train accuracies of the model
train_accuracies = {}

#A dictionary to srore the cross validation accuracies
cv_accuracies = {}

## **DEFINING THE MODEL**

Different **Sequential architectures** will be used to train the model. The model that performs very well on the cross validation set will be used to make predictions.

If the preferred model overfits, **regularization will be used to prevent the overfitting** so that the model can perform very well on the unseen data which is the test set

In [460]:
model = Sequential(
    [
        Dense(units=15, activation='relu'),
        Dense(units=8, activation='tanh'),
        Dense(units=10, activation='tanh'),
        Dense(units=1, activation='linear')
    ], name = 'Model_1'
)

## **Trains the neural network**

* This trains the neural network using the training data and the data specified below

* Trains the neural network using the X_train_scaled dataset

* Trains the training data using the models specified in the above code

* Validates each of the models using the cross validation dataset to select the best model for fine-tuning the model

* Specifies and imports the **BinaryCrossEntropy cost function** to be used to train the neural network. Since the output layer is a **logistic regression (binary layer)**, the from_logits parameter in the function will be set to true.

* The **from_logits set to True** shows that the last layer in the neural network is **multilable classification** hence tensorFlow uses a more error tolerating logistic **regression function as the activation function** for that layer, **reducing run-off errors**

* Specifies the **Adaptive Gradient Descent (Adam) optimizer** to be used during training to enable gradiesnt descent get to the local minimum faster which inturn speeds up the training process

* Specifies the **epochs (number of iterations)** during training for each of the models

* Verbose set to 1 or 2 shows an animated bar of the epoches during training

In [461]:
#Imports the BinaryCrossentropy from tensorflow.keras.losses
from tensorflow.keras.losses import BinaryCrossentropy

#Imports the Adaptive Gradient Descent (Adam) from tensorflow.kera.optimizers
from tensorflow.keras.optimizers import Adam

In [462]:
#Calls the compile method on the defines model
model.compile(
    loss = BinaryCrossentropy(from_logits=True),
    optimizer = Adam(0.001)
)

In [463]:
#Calls the fit method on the X_train_scaled data
model.fit(
    X_train_scaled, y_train,
    epochs = 200,
    verbose =  0
    )

In [464]:
#Use the trained model to predict the X_train scaled data
prediction_train = model.predict(X_train_scaled)

#Runs the predicted values through a sigmoid from since the output is binary. The output is the probability of the output being 0 or 1
prediction_prob_train = tf.nn.sigmoid(prediction_train).numpy()

#Converts each predicted output as 0 or 1 using the calculated probabilities
prediction_train = (prediction_prob_train >= 0.5).astype(int)

#Calculates the accuracy score
accuracy_score_train = accuracy_score(y_train, prediction_train)

print(f'The train accuracy is {accuracy_score_train}')

26/26 [==============================] - 0s 1ms/step
The train accuracy is 0.8626716604244694


In [465]:
#Use the trained model to predict the X_cv scaled data
prediction_cv = model.predict(X_cv_scaled)

#Runs the predicted values through a sigmoid since the output is binary. The output is the probability of the output being 0 or 1
prediction_prob_cv = tf.nn.sigmoid(prediction_cv).numpy()

#Converts each predicted output as 0 or 1using the calculated probabilities
prediction_cv = (prediction_prob_cv >= 0.5).astype(int)

#Calculates the accuracy score
accuracy_score_cv = accuracy_score(y_cv, prediction_cv)

print(f'The cv accuracy is {accuracy_score_cv}')

3/3 [==============================] - 0s 3ms/step
The cv accuracy is 0.7888888888888889


## **REGULARIZATION**

The model performs very well with an train accuracy being 87% and cross validation 82%

The selected model will be regularized to prevent overfitting therefore enabling the model to generalize pretty on unseen data that is the test set.

In [466]:
reg_model =  Sequential(
    [
        Dense(units=15, activation='relu', kernel_regularizer = tf.keras.regularizers.l2(10)),
        Dense(units=8, activation='tanh', kernel_regularizer = tf.keras.regularizers.l2(10)),
        Dense(units=10, activation='tanh', kernel_regularizer = tf.keras.regularizers.l2(10)),
        Dense(units=1, activation='linear')
    ], name = 'Regularized_Model'
)

In [467]:
reg_model.compile(
  loss = BinaryCrossentropy(from_logits=True),
  optimizer = Adam(0.001)
)

In [468]:
reg_model.fit(
    X_train_scaled, y_train,
    epochs = 200,
    verbose = 0
)

In [469]:
#Use the trained model to predict the X_train scaled data
prediction_train = model.predict(X_train_scaled)

#Runs the predicted values through a sigmoid from since the output is binary. The output is the probability of the output being 0 or 1
prediction_prob_train = tf.nn.sigmoid(prediction_train).numpy()

#Converts each predicted output as 0 or 1 using the calculated probabilities
prediction_train = (prediction_prob_train >= 0.5).astype(int)

#Calculates the accuracy score
accuracy_score_train = accuracy_score(y_train, prediction_train)

print(f'The train accuracy is {accuracy_score_train}')

26/26 [==============================] - 0s 1ms/step
The train accuracy is 0.8626716604244694


In [470]:
#Use the trained model to predict the X_cv scaled data
prediction_cv = model.predict(X_cv_scaled)

#Runs the predicted values through a sigmoid since the output is binary. The output is the probability of the output being 0 or 1
prediction_prob_cv = tf.nn.sigmoid(prediction_cv).numpy()

#Converts each predicted output as 0 or 1using the calculated probabilities
prediction_cv = (prediction_prob_cv >= 0.5).astype(int)

#Calculates the accuracy score
accuracy_score_cv = accuracy_score(y_cv, prediction_cv)

print(f'The cv accuracy is {accuracy_score_cv}')

3/3 [==============================] - 0s 3ms/step
The cv accuracy is 0.7888888888888889
